stdlib/builtins.pyi

```diff
@@ -1619,12 +1618,27 @@ class reversed(Iterator[_T], Generic[_T]):
     def __length_hint__(self) -> int: ...
 
 def repr(__obj: object) -> str: ...
+
+# See https://github.com/python/typeshed/pull/9141 on why we don't use
+# `SupportsRound` from `typing.pyi`
+
+class _SupportsRound1(Protocol[_T_co]):
+    def __round__(self) -> _T_co: ...
+
+class _SupportsRound2(Protocol[_T_co]):
+    def __round__(self, __ndigits: int) -> _T_co: ...
+
+# We need these overloads to replicate how stdlib works
+@overload
+def round(number: float, ndigits: None = ...) -> int: ...
+@overload
+def round(number: int, ndigits: SupportsIndex) -> int: ...
 @overload
-def round(number: SupportsRound[Any]) -> int: ...
+def round(number: float, ndigits: SupportsIndex) -> float: ...
 @overload
-def round(number: SupportsRound[Any], ndigits: None) -> int: ...
+def round(number: _SupportsRound1[_T], ndigits: None = ...) -> _T: ...
 @overload
-def round(number: SupportsRound[_T], ndigits: SupportsIndex) -> _T: ...
+def round(number: _SupportsRound2[_T], ndigits: SupportsIndex) -> _T: ...
```

Comment:
note that you can pass arbitrary types for both arguments of round, the arguments are passed through unmodified to __round__, so imho the second argument shouldn't be restricted to int or SupportsIndex
see https://github.com/python/cpython/blob/231d83b72435d61e929d6d62f2be7305e7b5b62b/Python/bltinmodule.c#L2305-L2331

Reply:
Technically, yes. But, again: this is quite against its main purpose: `round(1.55, 'a')` must be reported. This is a clear error.

I don't think people are basing their API around non-`int` arguments to `round`

Reply:
Agree with @sobolevn. Practicality beats purity here; it doesn't serve users well to allow a `str` argument (or whatever) to `ndigits`

Reply:
`round(1.55, 'a')` would be a type error because `float.__round__` doesn't allow a string. imho there are valid use-cases for non-int types, e.g. passing an object that is a symbolic variable (like how sympy has symbolic variables): `repr(round(Symbol('x'), Symbol('y'))) == 'round(x, y)'`. imho it should be more like:
```python
class _SupportsRound2(Protocol[_T, _R]): # add co/contravariance as necessary
    def __round__(self, __ndigits: _T) -> _R: ...
@overload
def round(__x: _SupportsRound2[_T, _R], ndigits: _T) -> _R: ...
```

Reply:
Interesting. Would the `Symbol` type in that example not have an `__index__` method @programmerjake? (I haven't really used sympy.)

Reply:
The [docs for `round()`](https://docs.python.org/3/library/functions.html#round) state:

> Any integer value is valid for _ndigits_ (positive, zero, or negative).

`SupportsIndex` reflects that quite well — it allows `int`, but also allows other alternative integer types that might exist, as long as they have an `__index__` method

Reply:
note, idk if sympy actually does that, i just want it to be valid to write a `Symbol` class that can do that. `Symbol.__index__` if it existed would probably return a symbolic expression, not an `int`.

Reply:
I'm sympathetic, but I still think we probably shouldn't support it since, according to the Python docs, `round()` just isn't meant to be used with a non-integer passed to the second argument :)

Reply:
> The [docs for `round()`](https://docs.python.org/3/library/functions.html#round) state:
> 
> > Any integer value is valid for _ndigits_ (positive, zero, or negative).

i interpret that to be for builtin self types only, since it's in the paragraph starting with "For the built-in types supporting round()". It states in a separate paragraph: "For a general Python object number, round delegates to number.__round__" -- imho this implies that `__round__`'s type should be used, rather than forcing integer-like types (SupportsIndex).

Reply:
I see what you mean, but it still says: "For a general Python object _number_" — implying `round()` is generally meant to be used with numeric types. Would you consider `Symbol` in your example to be a "numeric" type?

Reply:
I've just tested proposed `_SupportsRound2[_T, _T_co]` protocol. It does not work, tests fail:

```
test_cases/stdlib/builtins/check_round.py:13: error: Expression is of type "Any", not "float"  [assert-type]
test_cases/stdlib/builtins/check_round.py:13: error: Cannot infer type argument 1 of "round"  [misc]
test_cases/stdlib/builtins/check_round.py:14: error: Expression is of type "Any", not "float"  [assert-type]
test_cases/stdlib/builtins/check_round.py:14: error: Cannot infer type argument 1 of "round"  [misc]
test_cases/stdlib/builtins/check_round.py:15: error: Expression is of type "Any", not "float"  [assert-type]
test_cases/stdlib/builtins/check_round.py:15: error: Cannot infer type argument 1 of "round"  [misc]
test_cases/stdlib/builtins/check_round.py:16: error: Expression is of type "Any", not "float"  [assert-type]
test_cases/stdlib/builtins/check_round.py:16: error: Cannot infer type argument 1 of "round"  [misc]
test_cases/stdlib/builtins/check_round.py:46: error: Expression is of type "Any", not "str"  [assert-type]
test_cases/stdlib/builtins/check_round.py:46: error: Cannot infer type argument 1 of "round"  [misc]
```

Source:

```python
class _SupportsRound1(Protocol[_T_co]):
    def __round__(self) -> _T_co: ...

class _SupportsRound2(Protocol[_T, _T_co]):
    def __round__(self, __ndigits: _T) -> _T_co: ...

@overload
def round(number: _SupportsRound1[_T], ndigits: None = ...) -> _T: ...
@overload
def round(number: _SupportsRound2[_S, _T], ndigits: _S) -> _T: ...
```

Reply:
> I see what you mean, but it still says: "For a general Python object _number_" (emphasis mine) — implying `round()` is generally meant to be used with numeric types. Would you consider `Symbol` in your example to be a "numeric" type?

yes, because it represents an arbitrary-ish number. an example where round is used with a symbol: https://www.wolframalpha.com/input?i=round%28x%29-floor%28x%29

Reply:
> I've just tested proposed `_SupportsRound2[_T, _T_co]` protocol. It does not work, tests fail:

i tried some variations...it appears as though neither mypy nor pyright can correctly deduce the types, their type solvers are just not powerful enough...i guess I'll have to give up for now.